src/backend/columnar/columnar_customscan.c

```diff
@@ -525,3 +536,112 @@ ColumnarScan_ExplainCustomScan(CustomScanState *node, List *ancestors,
 							   chunkGroupsFiltered, es);
 	}
 }
+
+
+static Node *
+ExprListBindParams_mutator(Node *node, ParamListInfo paramLI)
+{
+	if (node == NULL)
+	{
+		return NULL;
+	}
+
+	/* copied from eval_const_expressions() */
+	if (IsA(node, Param))
```

Comment:
Instead of copying pg code, I made a hack in commit 9f64c12c7f821c0eccb92f1dd5e3fdb86722f5dc, what do you think ?

I think `qual` nodes other than `Param`s anyway will not be processed by `eval_const_expressions_mutator` since the fake `PlannerInfo` passed to `eval_const_expressions` wouldn't have enough info to process them.

------

Also, with commit 9f64c12c7f821c0eccb92f1dd5e3fdb86722f5dc, I was hoping to make chunk group filtering work for below query, but ..:

```sql
CREATE TABLE CASE_TBL (
  i integer,
  f double precision
) USING columnar;

SELECT alter_columnar_table_set('CASE_TBL', chunk_group_row_limit => 1000);

INSERT INTO CASE_TBL SELECT i % 2000, i FROM generate_series(1, 5000) i;

PREPARE prep_null_if (int) AS
SELECT * FROM CASE_TBL WHERE i > NULLIF($1, 2);

-- doesn't prune any chunk groups even in the first execution
explain analyze execute prep_null_if(3000);
```

Indeed, I realized that this is not related with prepared statements but because we still don't pass the actual `PlannerInfo` to  `eval_const_expressions`:

```sql
-- doesn't prune any chunk groups again
explain analyze SELECT * FROM CASE_TBL WHERE i > NULLIF(3000, 2);
```

I think we can either pass the original `PlannerInfo` to `eval_const_expressions` in the commit that I shared, or maybe we should be okay with such scenarios for now, but better to have some tests for them.

Reply:
Also made some testing with dropped columns and partitioned tables, all seem to work very well.
(Fell free to include the tests 9cddf8f08060d7057abcfac1fba73a99a4302d55)

Reply:
> Instead of copying pg code, I made a hack in commit [9f64c12](https://github.com/citusdata/citus/commit/9f64c12c7f821c0eccb92f1dd5e3fdb86722f5dc), what do you think @jeff-davis ?

That was actually my first version. It felt a little too hacky so I copied just the code for Param substitution, but on reflection I think you're right. It's better to just use the dummy structure.

It's possible we missed a case, or a future change will cause a crash when it looks at the bogus structure. But that seems like a fairly small risk.
 
> I think `qual` nodes other than `Param`s anyway will not be processed by `eval_const_expressions_mutator` since the fake `PlannerInfo` passed to `eval_const_expressions` wouldn't have enough info to process them.

It actually may do some work, because after the Params are replaced, more expressions might be folded to Consts. Consider `x = $1 + 17`.

> Also, with commit [9f64c12](https://github.com/citusdata/citus/commit/9f64c12c7f821c0eccb92f1dd5e3fdb86722f5dc), I was hoping to make chunk group filtering work for below scenario, but ..:

That seems to be a problem where `NULLIF` isn't const-simplified. But your change does work for other cases.
 
> I think we can either pass the original `PlannerInfo` to `eval_const_expressions` in the commit that I shared, or maybe we should be okay with such scenarios for now, but better to have some tests for them.

How would we do that? The PlannerInfo is available at plan time, but the ParamListInfo isn't available until execution time.

Reply:
> How would we do that? The PlannerInfo is available at plan time, but the ParamListInfo isn't available until execution time.

👍